src/xref2/strengthen.ml

```diff
@@ -20,21 +20,30 @@ open Delayed
 let rec signature :
     Cpath.module_ -> ?canonical:Cpath.module_ -> Signature.t -> Signature.t =
  fun prefix ?canonical sg ->
+  let sg', strengthened_modules = sig_items prefix ?canonical sg in
+  (* Format.eprintf "Invalidating modules: %a\n%!" (Format.pp_print_list Ident.fmt) strengthened_modules; *)
+  let substs =
+    List.fold_left
+      (fun s mid -> Subst.path_invalidate_module (mid :> Ident.path_module) s)
+      Subst.identity strengthened_modules
+  in
+  Subst.signature substs sg'
+
+and sig_items prefix ?canonical sg =
   let open Signature in
-  let items, strengthened_modules =
+  let items, ids =
     List.fold_left
       (fun (items, s) item ->
         match item with
-        | Module (id, r, m) -> (
+        | Module (id, r, m) ->
             let name = Ident.Name.module_ id in
             let canonical =
               match canonical with
               | Some p -> Some (`Dot (p, name))
               | None -> None
             in
-            match module_ ?canonical (`Dot (prefix, name)) (get m) with
-            | None -> (item :: items, s)
-            | Some m' -> (Module (id, r, put (fun () -> m')) :: items, id :: s))
+            let m' = module_ ?canonical (`Dot (prefix, name)) (get m) in
+            (Module (id, r, put (fun () -> m')) :: items, id :: s)
```

Comment:
The whole computation of `m'` could be delayed (it's not used otherwise)

Reply:
Good spot!